File: 797. All Paths From Source to Target/Solution.java
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

class Solution {
    List<List<Integer>> result = new LinkedList<>();

    public List<List<Integer>> allPathsSourceTarget(int[][] graph) {
        Map<Integer, int[]> graphMap = new HashMap<>();
        for (int i = 0; i < graph.length; i++) {
            graphMap.put(i, graph[i]);
        }
        testRoad(0, graph.length - 1, new LinkedList<>(), graphMap);
        return result;
    }

    public void testRoad(int root, int end, List<Integer> road, Map<Integer, int[]> graphMap) {
        road.add(root);
        if (root == end) {
            result.add(road);
            return;
        } else if (graphMap.get(root).length == 0)
            return;
        else {
            for (int integer : graphMap.get(root)) {
                testRoad(integer, end, new LinkedList<>(road), graphMap);
            }
        }
    }
}

File: 797. All Paths From Source to Target/Solution.py
from collections import defaultdict
from typing import List


class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        result = []
        graphMap = defaultdict(list)
        for i in range(len(graph)):
            graphMap[i] = graph[i]

        def test_road(root: int, end: int, road: List[int]):
            nonlocal graphMap, result
            road.append(root)
            if (root == end):
                result.append(road)
                return
            elif (len(graphMap[root]) == 0):
                return
            else:
                for child in graphMap[root]:
                    test_road(child, end, road.copy())

        test_road(0, len(graph)-1, [])
        return result
